aix: improve AIX compatibility for Node.js use

andrewlow opened the pull request:
As we work through the changes to Node required to support
PowerPC platforms, we have had to introduce a changes to the
libuv component to support the AIX operating system.

These are the initial enabling changes required.

andrewlow committed:
aix: improve AIX compatibility for Node.js use

As we work through the changes to Node required to support
PowerPC platforms, we have had to introduce a changes to the
libuv component to support the AIX operating system.

These are the initial enabling changes required.

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

The following commiters were not found in the CLA:
- Andrew Low

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

andrewlow commented:
To build on AIX I'm using gcc 4.6.1

The gyp flock support isn't working, so I've worked around that using `export LINK=g++`

saghul commented:
Wow, that's a big one :-) I don't have an AIX system to test on, but I'll take a look. Thanks!

andrewlow commented:
I build Node.js on AIX here: http://v8ppc.osuosl.org:8080/view/Node/ -- not quite the same code, but close.

Also look here: https://github.com/andrewlow/node in both the v0.10.26-release-ppc and v0.11.12-release-ppc branches for more detailed commit logs on the changes.

I can create a libuv build on the same system if needed. 

Files unique to AIX (should have no impact to other platforms)
- include/uv-aix.h
- src/unix/aix.c

Files with #ifdef gated changes
- include/uv-unix.h
- src/unix/core.c
- src/unix/fs.c
- src/unix/internal.h
- src/unix/tty.c
- src/unix/udp.c
- test/test-fs.c

You might need to look at uv.gyp and if I broke the _automake based build_ I'll need some pointers.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,244 @@
+#undef reqevents
+#undef rtnevents
+#undef events
+#undef revents
```
What's the reasoning behind this define / undefine maneuvers?

andrewlow replied:
On the surface it looks like an attempt to avoid a collision with some symbols. However this is a good question. The original commit that introduced this change can be found here: https://github.com/andrewlow/node/commit/fa1b515d7ef41337ff6e27266a11ce9995fe6a7f 

Removing the #define / #undef seems to have no negative impact to the code. I'll remove from the commit.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,244 @@
+
+  if (loop->backend_fd == -1)
+    return -1;
+
```
Extra newline.

andrewlow replied:
Github isn't picking up on my changes here - or I missed which newline you consider to be 'extra'.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,244 @@
+      pc.cmd = PS_ADD;
+      if (pollset_ctl(loop->backend_fd, &pc, 1)) {
+        if (errno != EINVAL) {
+          fprintf(stderr, "Failed to add file descriptor %d to pollset: %d, aborting at %d...\n", pc.fd, errno, __LINE__);
```
We generally don't print to stdout / stderr on errors. Add an assert / abort if the error is unrecoverable, but try to be gentle in case of failure ;-)

andrewlow replied:
I'm going to have to think about this a bit. A number of the fprintf's are immediately followed by an abort() and serve to provide the user better diagnostic information. Changing those into comments may be the simple solution that helps people who run into issues.

Some of them are informational, the operation is indicated as failing (gracefully) but an informational message is provided. I'm tempted to change these into some form of assert.

Looking around the code for other examples of handling errors, I came across uv__stream_recv_cmsg() in https://github.com/joyent/libuv/blob/master/src/unix/stream.c

```
      if (cmsg->cmsg_type != SCM_RIGHTS) {
      fprintf(stderr, "ignoring non-SCM_RIGHTS ancillary data: %d\n",
          cmsg->cmsg_type);
      continue;
    }
```

which seems to be providing informational output and continuing.

I may need more guidance on when information messages are ok.

saghul replied:
I'd say they are not OK here. The backtrace will tell you what went wrong it it aborts. You may add an `assert` so debug builds see more info though.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+    /* If it is a link, the resolved filename is again a relative path. make it absolute. */
+    else
+      (void) snprintf(buffer, PATH_MAX-1, "%s/%s", dirname(symlink), temp_buffer);
   *size = strlen(buffer);
```
Please indent the following lines, for readability.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+
+  if(argv == NULL)
+    return -1;
+  if(argv[0] == NULL)
```
Compact into a single if?

saghul replied:
Also, we need to return a UV return code here, which on Unix is the negative errno. Maybe -EINVAL?

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+    /* This is always a symlink, resolve it. */
+    res = readlink(cwd, cwdl, sizeof(cwdl) - 1);
+    if (res < 0)
+      return res;
```
return -errno;

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+
+    res = readlink(symlink, temp_buffer, PATH_MAX-1);
+    if(res < 0)
+    strcpy(buffer, symlink);
```
Indent.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+    if(res < 0)
+    strcpy(buffer, symlink);
+    else
+    (void) snprintf(buffer, PATH_MAX-1, "%s/%s", dirname(symlink), temp_buffer);
```
Indent.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+    /* This is always a symlink, resolve it. */
+    res = readlink(cwd, cwdl, sizeof(cwdl) - 1);
+    if (res < 0)
+      return res;
```
return -errno;

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+
+    res = readlink(symlink, temp_buffer, PATH_MAX-1);
+    if(res < 0)
+    strcpy(buffer, symlink);
```
Indent.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+    if(res < 0)
+    strcpy(buffer, symlink);
+    else
+    (void) snprintf(buffer, PATH_MAX-1, "%s/%s", dirname(symlink), temp_buffer);
```
Indent.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+    struct stat statstruct;
+
+    /* Get the paths. */
+    char *path=getenv("PATH");
```
spaces between = sign.

saghul replied:
Should we check if path is NULL? Not sure how things would go if PATH is unset. Who would do that though? :-)

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+    (void) snprintf(cwd, PATH_MAX-1, "/proc/%lu/cwd", (unsigned long) getpid());
+    res = readlink(cwd, cwdl, sizeof(cwdl) - 1);
+    if (res < 0)
+      return res;
```
return -errno;

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+          /* A path which is not current directory. */
+          (void) snprintf(symlink, PATH_MAX-1, "%s%s/%s", cwdl, relative+1, ps.pr_fname);
+        else
+        (void) snprintf(symlink, PATH_MAX-1, "%s%s", cwdl, ps.pr_fname);
```
Indent.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +320,144 @@ int uv_exepath(char* buffer, size_t* size) {
+      }
+      token = strtok(NULL, ":");
+    }
+    /* fprintf(stderr, "Yet to implement, could not find %s in %s\n", ps.pr_fname, path ); */
```
Remove leftover comment.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +494,402 @@ void uv_loadavg(double avg[3]) {
 }
 
 
+char *rawname(char *cp) {
```
Can these helper functions be static?

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +494,402 @@ void uv_loadavg(double avg[3]) {
+  int rc = 0;
+
+  if(stat(filename, &statbuf) < 0) {
+    fprintf(stderr, "file_is_a_directory - Error stating file: %s\n", strerror(errno));
```
Remove fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +494,402 @@ void uv_loadavg(double avg[3]) {
+
+  if(stat(filename, &statbuf) < 0) {
+    fprintf(stderr, "file_is_a_directory - Error stating file: %s\n", strerror(errno));
+    return(errno);
```
No need for parentheses. Maybe use -errno?

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +494,402 @@ void uv_loadavg(double avg[3]) {
+    return(errno);
+  }
+  if(statbuf.st_type == VDIR) {
+    rc=1;
```
Spacing.

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +897,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+    return rc;
+
+  /* Setup/Initialize all the libuv routines */
+  uv__handle_start(handle); /* FIXME shouldn't start automatically */
```
Bogus comment, we are in the start method :-)

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,15 @@ static ssize_t uv__fs_futime(uv_fs_t* req) {
 static ssize_t uv__fs_read(uv_fs_t* req) {
   ssize_t result;
 
+#if defined(_AIX)
```
Why not let read fail?

andrewlow replied:
Original change here https://github.com/andrewlow/node/commit/8f0861197b8d70a1b6acc439d32f426832cc81e4

AIX doesn't fail like Linux does for a directory read. This code causes AIX to behave like Linux.

saghul replied:
Wow. Out of curiosity, what does read() return for a directory?

andrewlow replied:
Now I'm doubting my previous comment as I've not been able to create a simple test case to show what was happening. I'll dig some more to figure out why we created this line.

andrewlow replied:
The following code

```
#include <unistd.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <stdio.h>


void format(char *str, int size)
{
        char temp[128];
        char *ptr = temp;
        int i = 0;
        for(i=0;i<size-1;)
        {
                while(*str == 0) {++str; i++;}
                ++str; ++str; i++; i++;
                while(*str != 0) { *ptr++ = *str++; i++;}
                *ptr = '\0';
                fprintf(stderr, "%s\n", temp);
                ptr = temp;
        }
}

int main()
{
        char buffer[128];
        int fd = open(".", O_RDONLY);
        if(fd == 0){fprintf(stderr, "Error opening directory: %d\n", errno); exit(1);}
        memset(buffer, 0, 128);
        int size = read(fd, buffer, 128);
        if(size < 0) {fprintf(stderr, "Error reading directory : %d.\n", errno); exit(1);}
        format(buffer, 128);
}
```

Will print the contents of the current working directory on AIX, and produce "Error reading directory : 21." on Linux Intel.

saghul commented on src/unix/tty.c:
```diff
@@ -191,6 +194,9 @@ uv_handle_type uv_guess_handle(uv_file file) {
 
   len = sizeof(sa);
+#if defined (_AIX)
+  memset(&sa, 0, sizeof(sa));
```
Why is this needed?

andrewlow replied:
https://github.com/andrewlow/node/issues/12

AIX can fail getsockname and NOT modify the buffer, however the memset may be overly cautious.

saghul replied:
AFAIS the buffer is not used in case of failure, so you could remove it.

saghul commented on uv.gyp:
```diff
@@ -135,6 +135,7 @@
             'include/uv-sunos.h',
             'include/uv-darwin.h',
             'include/uv-bsd.h',
+            'include/uv-aix.h',
```
You'll need to add this to the autoconf build system as well.

andrewlow replied:
~~The autoconf on the AIX development machine(s) I have access to is horribly busted.  Not sure how to proceed here ;(~~

Ick. I've managed to fix my AIX machine up and have autoconf working now.

saghul commented:
Did a quick first pass and left comments here and there.

indutny commented on src/unix/aix.c:
```diff
@@ -39,12 +39,244 @@
+
+    if (nfds == -1) {
+      if (errno != EINTR) {
+printf("pollset_poll() failure with errno: %d\n", errno);
```
Wrong indent.

andrewlow replied:
That printf looks like it shouldn't be there and is left-over debug code.

indutny commented on src/unix/aix.c:
```diff
@@ -71,15 +309,8 @@ int uv_exepath(char* buffer, size_t* size) {
-  cwdl[res] = '\0';
-
   (void) snprintf(pp, sizeof(pp), "/proc/%lu/psinfo", (unsigned long) getpid());
+
```
Unnecessary change

andrewlow commented:
Give me some time to revisit this again guys. I've released a patch that addresses at least some of the comments, but need to make another sweep to finish the clean-up. I'll also squash the commit(s) down to a single one when I'm done.

Some feedback on the use (or non-use) of fprintf would be helpful.

andrewlow commented:
At this point, I believe I've responded to / addressed all comments on the commit. Please let me know next steps.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+  int count;
+  int nfds;
+  int i;
+  int qry_rc;
```
Please rename this to 'r' or 'rc'.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+        qry_rc = pollset_query(loop->backend_fd, &pqry);
+        switch (qry_rc) {
+        case -1: 
+          fprintf(stderr, "Failed to add file descriptor %d to pollset: %d (EINVAL)\n", pc.fd, EINVAL);
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+          fprintf(stderr, "Pollset does not contain file descriptor %d, aborting at %d...\n", pc.fd, __LINE__);
+          abort();
+        }
+        /* If we got here then the pollset already contained the file descriptor even though
```
Maybe use default: here?

andrewlow replied:
Unless you have strong feelings about it, I think we'll leave it as written.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+      }
+    }
+    if (w->events != 0 || add_failed) {
+      /* Modify, potentially removing events -- need to delete then add */
```
Merge 2 comments into one.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+       * lazily remove events, squelching them in the mean time. */
+      pc.cmd = PS_DELETE;
+      if (pollset_ctl(loop->backend_fd, &pc, 1)) {
+        fprintf(stderr, "Failed to delete file descriptor %d from pollset: %d, aborting at %d...\n", pc.fd, errno, __LINE__);
```
Remove fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+      }
+      pc.cmd = PS_ADD;
+      if (pollset_ctl(loop->backend_fd, &pc, 1)) {
+        fprintf(stderr, "Failed to add file descriptor %d to pollset: %d, aborting at %d...\n", pc.fd, errno, __LINE__);
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+        continue;
+      }
+
+      w->cb(loop, w, pe->revents);
```
Any chance revents is 0, or contains events we don't care about? (check linux-core.c, some filtering is done there, does a similar thing also apply here?)

andrewlow replied:
I'm not sure. The solaris implementation of this function (src/unix/sunos.c) doesn't contain the same filtering logic as linux-core.c - this makes me hesitant to just add it to AIX "because".

Is there a test case that you are aware of that pushes on this code and exposes the 'extra events'? 

For now, I'd rather leave it as is and deal with issues as we go forwards.

saghul replied:
Ok.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,147 @@ int uv_exepath(char* buffer, size_t* size) {
+     * and see which one succeeds. Exit on first match. */
+    while(token != NULL) {
+      if(token[0] == '.') {
+
```
Extra newline.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,147 @@ int uv_exepath(char* buffer, size_t* size) {
+        else
+          (void) snprintf(symlink, PATH_MAX-1, "%s%s", cwdl, ps.pr_fname);
+        if(stat(symlink, &statstruct) != -1) {
+
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,147 @@ int uv_exepath(char* buffer, size_t* size) {
+     * make it absolute. 
+     */
+    else
+      (void) snprintf(buffer, PATH_MAX-1, "%s/%s", dirname(symlink), temp_buffer);
```
Why the (void)? We seem to only do it in AIX.

andrewlow replied:
Style - but I've removed to be consistent.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+  char *obj, *stub;
+
+  if((p = (struct vmount *)malloc(siz)) == (struct vmount *)NULL) {
+    fprintf(stderr, "is_ahafs_mounted - malloc: %s\n", strerror(errno));
```
fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+  const char *dev = "/aha";
+  char *obj, *stub;
+
+  if((p = (struct vmount *)malloc(siz)) == (struct vmount *)NULL) {
```
Unnecessary cast. Also, please do it in 2 steps.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+  /* Retrieve all mounted filesystems */
+  while(1) {
+    rv = mntctl(MCTL_QUERY,siz,(char*)p);
+    if(rv>0) {
```
spaces between operator

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+    if(rv>0) {
+      break;        /* It worked */
+    }
+    if(rv<0) {
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+      break;        /* It worked */
+    }
+    if(rv<0) {
+      fprintf(stderr, "is_ahafs_mounted - mntctl: %s\n", strerror(errno));
```
fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+    /* Use a bigger size buffer to hold all mounted filesystems */
+    siz = sizeof(struct vmount)*sizeMultiplier * i;
+    free(p);
+    p = (struct vmount *)malloc(siz);
```
Unneeded cast.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+    free(p);
+    p = (struct vmount *)malloc(siz);
+
+    if(p == (struct vmount *)NULL) {
```
Unneded cast.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+    p = (struct vmount *)malloc(siz);
+
+    if(p == (struct vmount *)NULL) {
+      fprintf(stderr, "is_ahafs_mounted - realloc: %s\n", strerror(errno));
```
fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+  }
+
+  /* Look for dev in filesystems mount info */
+  for(vmt = (struct vmount *)p, i = 0; i < rv; i++) {
```
Unneeded cast.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +487,403 @@ void uv_loadavg(double avg[3]) {
+  }
+
+  /* Error Message*/
+  fprintf(stderr, "/aha is required for monitoring filesystem changes\n");
```
fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +891,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+  int fileIsDirectory = 0; /* NO==0, YES==1, error otherwise*/
+
+  if (!uv__is_active(handle))
+    return -EINVAL;
```
return 0 here. I'll fix it on the rest of the platforms, it should be this way.

saghul commented on src/unix/internal.h:
```diff
@@ -42,6 +42,16 @@
 #endif /* __sun */
 
+#if defined(_AIX)
+#define reqevents events
```
Are these needed? IIRC you removed them from somewhere else.

andrewlow replied:
These are needed as the AIX build still make use of src/unix/core.c and the following functions:
- `uv__io_init()`
- `uv__io_start()`
- `uv__io_stop()`

make use of the `reqevents` field. By doing this `#ifdef` remapping of the AIX header file we achieve compatibility.

saghul replied:
I missed your reply here. So, why is the name mangling required again? Is this the same problem? https://bugs.ruby-lang.org/issues/8174

saghul commented on src/unix/tty.c:
```diff
@@ -204,6 +207,18 @@ uv_handle_type uv_guess_handle(uv_file file) {
+  if(len == 0) {
+    len = sizeof(sa);
+    if (getpeername(file, &sa, &len) == -1) {
+      if(file == 0 || file == 1 || file == 2)
```
Not sure if this is a good idea. What if fd 0 was closed and then assigned to a TCP socket?

andrewlow replied:
This is a limitation we need to (unfortunately) accept on AIX. 

These lines are a combination of two commits:
https://github.com/andrewlow/node/commit/c1d565ba92ab1eb3c6950904653b8fb19b938405
https://github.com/andrewlow/node/commit/0edef8289a44fd67c631cdb15bc9060bee68c522

AIX lacks socket introspection support for AF_UNIX sockets, however the standard I/O pipes also look just like a socket to this code so we have to hope it is the usual case.

saghul commented on src/unix/tty.c:
```diff
@@ -204,6 +207,18 @@ uv_handle_type uv_guess_handle(uv_file file) {
+    if(sa.sa_family == AF_UNIX)
+      return UV_NAMED_PIPE;
+  }
+#endif /* defined(_AIX) */
```
Can you explain why this is needed?

txdv replied:
Why is there sometimes a space after an if and sometimes not?

andrewlow replied:
Code formatting is just wrong (missing space) - thanks @txdv 

@saghul please refer to the commit https://github.com/andrewlow/node/commit/c1d565ba92ab1eb3c6950904653b8fb19b938405 which is a work-around for the lack of AF_UNIX socket introspection ability on AIX.

This is a trade-off between correctness and working code in the common cases. We are forced to guess as AIX simply doesn't have the same capability as other Unix in this area.

saghul commented:
Round 2 done.

andrewlow commented:
I've responded to all of the round 2 comments. Thank you @saghul for your detailed review.

andrewlow commented:
Bump - could someone comment on next steps here? If you guys are just busy, that's cool - I just want to know the status.

indutny commented:
Heya @andrewlow ! @saghul is on a trip, I am too :) Anyway, it would be better if @saghul would continue reviewing it, because I'm not really into the current state of it.

andrewlow commented:
@indutny thanks for the ultra-fast reply. I'm happy to wait for @saghul to get back from his trip.

saghul commented:
Sorry for the delay! I'll be back next week and give it another review
round.
On Jun 4, 2014 3:16 PM, "Andrew Low" notifications@github.com wrote:

> @indutny https://github.com/indutny thanks for the ultra-fast reply.
> I'm happy to wait for @saghul https://github.com/saghul to get back
> from his trip.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joyent/libuv/pull/1256#issuecomment-45088560.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+        rc = pollset_query(loop->backend_fd, &pqry);
+        switch (rc) {
+        case -1: 
+          assert(0 && "Failed to add file descriptor (pc.fd) to pollset");
```
You'd never see the second assert, can you combine the test of both?

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,237 @@
+          assert(0 && "Failed to query pollset for file descriptor");
+          abort();
+        case 0:
+          assert(0 && "Failed to add file descriptor (pc.fd) to pollset");
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,145 @@ int uv_exepath(char* buffer, size_t* size) {
+     * outbut buffer. 
+     */
+    if (res < 0)
+      strcpy(buffer, symlink);
```
Check for error.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,145 @@ int uv_exepath(char* buffer, size_t* size) {
+    if (res < 0)
+      strcpy(buffer, symlink);
+
+    /* If it is a link, the resolved filename is again a relative path,
```
If you are adding comments, please but braces on the if, makes it more readable.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,145 @@ int uv_exepath(char* buffer, size_t* size) {
+
+    res = readlink(symlink, temp_buffer, PATH_MAX-1);
+    if (res < 0)
+      strcpy(buffer, symlink);
```
Check for error.

saghul replied:
Disregard these, my brain was reading "strdup".

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,145 @@ int uv_exepath(char* buffer, size_t* size) {
+
+    res = readlink(symlink, temp_buffer, PATH_MAX-1);
+    if (res < 0)
+      strcpy(buffer, symlink);
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,145 @@ int uv_exepath(char* buffer, size_t* size) {
+        if (stat(symlink, &statstruct) != -1) {
+          res = readlink(symlink, temp_buffer, PATH_MAX-1);
+          if (res < 0)
+            strcpy(buffer, symlink);
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +310,145 @@ int uv_exepath(char* buffer, size_t* size) {
+      strcpy(buffer, symlink);
+    else
+      snprintf(buffer, PATH_MAX-1, "%s/%s", dirname(symlink), temp_buffer);
+    *size = strlen(buffer);
```
Need to check if the supplied buffer is big enough.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
 }
 
 
+static char *rawname(char *cp) {
```
prefix internal functions with `uv__`.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+}
+
+
+/* NAME:    file_is_a_directory
```
PLease reword these comments to match our style in uv.h.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+ *   0 - NO
+ *   -errno on failure 
+ *  TODO: most callers of this don't check the error case
+ */
```
Prefix with `uv__`.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+ *   0 - AHAFS is mounted.
+ *  -1 - AHAFS is not mounted
+ */
+static int is_ahafs_mounted(void){
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+ * RETURNS:
+ *  Return code from mkdir call
+ */
+static int _mkdir(const char *dir) {
```
Prefix. Also, call it `uv__makedir_p` or something alike.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+ * RETURNS:
+ *  Return code from mkdir call
+ */
+static int mk_subdirs(const char *filename) {
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+  rc = _mkdir(cmd);
+
+  if (rc == -1 && errno != EEXIST){
+    fprintf(stderr, "mk_subdirs error: %s\n", strerror(errno));
```
Remove fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+ *   0 - Successfull
+ *   Error otherwise
+ */
+static int setup_ahafs(const char* filename, int *fd) {
```
Prefix.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+  }
+
+  if ((strlen(monFile) + strlen(filename) + 5) > PATH_MAX) {
+    fprintf(stderr, "Error: Cannot monitor object, path name too long\n");
```
No fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+
+  if ((strlen(monFile) + strlen(filename) + 5) > PATH_MAX) {
+    fprintf(stderr, "Error: Cannot monitor object, path name too long\n");
+    return(ENAMETOOLONG);
```
No need for parens. Also, if this error is supposed to be returned to the user, use negative errno.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+  /* Open the monitor file, creating it if necessary */
+  *fd = open(monFile, O_CREAT|O_RDWR);
+  if (*fd < 0) {
+    fprintf(stderr, "setup_ahafs - Error opening monitor file: %s\n", strerror(errno));
```
No fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+
+  rc = write(*fd, monFileWrStr, strlen(monFileWrStr)+1);
+  if (rc < 0) {
+    fprintf(stderr, "setup_ahafs - Error writing to monitor file: %s\n", strerror(errno));
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+  rc = write(*fd, monFileWrStr, strlen(monFileWrStr)+1);
+  if (rc < 0) {
+    fprintf(stderr, "setup_ahafs - Error writing to monitor file: %s\n", strerror(errno));
+    return(errno);
```
No parens, use negative errno.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+ * RETURNS:
+ *      Total number of lines skipped
+ */
+static int skip_lines(char **p, int n) {
```
Prefix

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+    n--;
+    lines++;
+  }
+  return(lines);
```
Parens.

andrewlow replied:
Sorry, been staring at this a bit and I can't figure out what the comment is asking to do here.

saghul replied:
Just `return lines;`, instead of `return(lines);`.

andrewlow replied:
Ok - gotcha.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+ *        0 - No corrective action needed.
+ *       -1 - Unrecoverable error in parsing
+ */
+static int parse_data(char *buf, int err, int *events, uv_fs_event_t* handle) {
```
Prefix.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+  for(i=0; i<PATH_MAX; i++) {
+    filename[i] = 0;
+  }
+  i=0;
```
Space around assignment.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+
+  /* Check for BUF_WRAP */
+  if (strncmp(buf, "BUF_WRAP", strlen("BUF_WRAP")) == 0) {
+    printf("Buffer wrap detected, Some event occurrences lost!\n");
```
No printf.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +485,404 @@ void uv_loadavg(double avg[3]) {
+  bytes = pread(event_watch->fd, resultData, RDWR_BUF_SIZE, 0);
+
+  if (bytes < 0)
+    fprintf(stderr, "uv__ahafs_event - Error reading monitor file: %s\n", strerror(errno));
```
No fprintf.

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +890,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
 
 int uv_fs_event_stop(uv_fs_event_t* handle) {
-  return -ENOSYS;
+  int fileIsDirectory = 0; /* NO==0, YES==1, error otherwise*/
```
no camel case var names.

saghul commented on src/unix/tty.c:
```diff
@@ -204,6 +207,18 @@ uv_handle_type uv_guess_handle(uv_file file) {
     if (sa.sa_family == AF_UNIX)
       return UV_NAMED_PIPE;
   }
+#if defined (_AIX)
```
Not sure if I asked before, but why is this needed on AIX? Also, what if a file is redirected as stdin, wouldn't we be reporting that it's a named pipe?

andrewlow replied:
There are two commits related to this change.
- https://github.com/andrewlow/node/commit/1e742f3db7eb6f290133754de2da835fbd965e23
- https://github.com/andrewlow/node/commit/5e8ef30c8278411903a7def845688829c0c4b162

The first is an AIX work-around as it lacks socket introspection support for AF_UNIX sockets.
The second is specifically identifying 0,1,2 files as stdio pipes - then reporting them as UV_NAMED_PIPE - which is what you're pointing out.

Is there a return value that would be more appropriate for stdio pipes?

In testing Node - this seemed to do the right thing for the tests. I can imagine scenarios where someone has closed stdout and then re-allocated that special numbered handle to point at something else. This would of course break the logic here, but we're already doing a best guess work-around. AIX simply doesn't support what we need.

saghul replied:
> The second is specifically identifying 0,1,2 files as stdio pipes - then reporting them as UV_NAMED_PIPE - which is what you're pointing out.
> 
> Is there a return value that would be more appropriate for stdio pipes?
> 
> In testing Node - this seemed to do the right thing for the tests. I can imagine scenarios where someone has closed stdout and then re-allocated that special numbered handle to point at something else. This would of course break the logic here, but we're already doing a best guess work-around. AIX simply doesn't support what we need.

Well, if AIX doesn't support it then we need to return UV_HANDLE_UNKNOWN. I'd rather return no result than a misleading one.

andrewlow replied:
Are you suggesting removing the `#if defined (_AIX)` entirely? That would break some Node tests.

This code is assuming that AF_UNIX pipes are named pipes, and doing a special case for 0,1,2 - otherwise it is still returning UV_UNKNOWN_HANDLE.

saghul commented:
Sorry for the delay :-S Just finished another round of comments. Mostly style stuff and I think there is a bounds check problem in uv_exepath.

andrewlow commented:
Again, appreciate the detailed review - and prompt replies to my most recent queries.

Delay wasn't a big deal, I was on vacation for a bit (from the keyboard too!) and introduced additional delay. I think a new sub-goal is to make this not only one of the larger pull requests, but one of the longest in duration :)

I believe at this point I've addressed all the comments, done some additional clean-up related to the comments (I fixed all the returns to be paren free, corrected all the comments).

There is the outstanding thread on src/unix/tty.c, but nothing else as far as I know.

saghul commented:
> I believe at this point I've addressed all the comments, done some additional clean-up related to the comments (I fixed all the returns to be paren free, corrected all the comments).

I haven't reviewed it in detail yet, but things look good at a first glance :-)

> There is the outstanding thread on src/unix/tty.c, but nothing else as far as I know.

So, how about this: drop those changes for now, we land this, and treat that as a separate issue. We plan to release libuv 0.12 soon and I'd like to have these changes in. This little detail can be handled as an issue, bugfixes are ok. Maybe we can even use some other approach like `stat()`-ing the fd instead of doing the getpeername + guessing dance.

andrewlow commented:
src/unix/tty.c changes for AIX dropped from the pull request.

Hopefully I didn't miss your 0.12 window.

saghul commented:
> src/unix/tty.c changes for AIX dropped from the pull request.
> 
> Hopefully I didn't miss your 0.12 window.

Thanks! Nope, there are still some pending PRs, don't worry :-) I'll give this (hopefully) a final round tonight.

txdv commented:
You are in the milestone target

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +308,160 @@ int uv_exepath(char* buffer, size_t* size) {
+    } else {
+      assert(*size > (strlen(symlink) + 1 + strlen(temp_buffer)));
+      snprintf(buffer, *size-1, "%s/%s", dirname(symlink), temp_buffer);
+    *size = strlen(buffer);
```
style: fix indent

saghul replied:
Shouldn't that go outside the else?

andrewlow replied:
Good catch. No idea what finger fumble caused that. Bracket went in on the wrong line. Fixed

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +308,160 @@ int uv_exepath(char* buffer, size_t* size) {
+    char *path = getenv("PATH");
+
+    /* Get a local copy. */
+    strcpy(clonedpath, path);
```
Check length of path. I assume that each path in PATH could potentially be of almost PATH_MAX chars, so you'd overflow.

andrewlow replied:
clonedpath is declared as PATH_MAX in size - it should fit any PATH.

I'll add an assert.

andrewlow replied:
Oh.. you're correct. PATH_MAX is per path entry, sorry.

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +308,160 @@ int uv_exepath(char* buffer, size_t* size) {
+      }
+      token = strtok(NULL, ":");
+    }
+    return -EINVAL;
```
A comment on how we can arrive here would be nice ;-)

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  char *dp = rindex(cp, '/');
+
+  if (dp == 0)
+    return (0);
```
style: return 0

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  *dp = '/';
+  strcat(rawbuf, "/r");
+  strcat(rawbuf, dp+1);
+  return (rawbuf);
```
style: return rawbuf

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  struct stat statbuf;
+  int rc = 0;
+
+  if (stat(filename, &statbuf) < 0) {
```
Drop parens for single statement if

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  if (stat(filename, &statbuf) < 0) {
+    return -errno;
+  }
+  if (statbuf.st_type == VDIR) {
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+    return -errno;
+  }
+  if (statbuf.st_type == VDIR) {
+    rc = 1;
```
No need to store rc, just return 1

saghul replied:
Actually, this function is a bit weird: it returns -errno for failure, 1 for success and 0 for other types of failure. Since it's called `uv__file_is_a_directory` i'd go for returning 1 on sucess and 0 on failure, swallowing the error.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+
+/* 
+ * Check whether AHAFS is mounted.
+ * Returns 0 if AHAFS is mounted, -1 if not
```
Try to be consistent with the above function. I don't care much about what model you pick since they are internal functions, but make it consistent.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  p = malloc(siz);
+  if (p == NULL) {
+    assert(0 && "malloc failed (errno)");
+    return -1;
```
If you go with < 0 failure, 0 success, then return -errno and let the caller deal with it.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  }
+
+  /* Error - aha not found */
+  assert(0 && "/aha is required for monitoring filesystem changes");
```
This doesn't look like a fatal error to me, why the assert? If /aha is not there `uv_fs_event_start` should return UV_ENOSYS.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  rc = uv__makedir_p(cmd);
+
+  if (rc == -1 && errno != EEXIST){
+    assert(0 && "uv__make_subdirs_p error");
```
Can't we just return the error?

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+    sprintf(mon_file, "/aha/fs/modFile.monFactory");
+
+  if ((strlen(mon_file) + strlen(filename) + 5) > PATH_MAX) {
+    assert(0 && "Error: Cannot monitor object, path name too long\n");
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  /* Open the monitor file, creating it if necessary */
+  *fd = open(mon_file, O_CREAT|O_RDWR);
+  if (*fd < 0) {
+    assert(0 && "uv__setup_ahafs - Error opening monitor file");
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+
+  rc = write(*fd, mon_file_write_string, strlen(mon_file_write_string)+1);
+  if (rc < 0) {
+    assert(0 && "uv__setup_ahafs - Error writing to monitor file");
```
Ditto.

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  for(i = 0; i < PATH_MAX; i++) {
+    filename[i] = 0;
+  }
+  i=0;
```
style: spaces around '='

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  else
+    return -1;
+
+  return rc;
```
You don't seem to be storing rc anywhere, just return 0 here?

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+/* This is the internal callback */
+static void uv__ahafs_event(uv_loop_t* loop, uv__io_t* event_watch, unsigned int fflags) {
+  char   result_data[RDWR_BUF_SIZE];
+  int bytes, rc=0;
```
style: spaces around assignment

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +498,390 @@ void uv_loadavg(double avg[3]) {
+  int bytes, rc=0;
+  uv_fs_event_t* handle;
+  int events=0;
+  int  i=0;
```
ditto x2

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +889,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
                       const char* filename,
                       unsigned int flags) {
-  return -ENOSYS;
+  int  fd, rc, i=0, res=0;
```
style: spaces

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +889,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+  }
+
+  if (uv__is_ahafs_mounted() == -1) {
+    /* Set errno or return whatever error is appropriate */
```
return UV_ENOSYS or UV_ENOTSUP here

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +889,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+  }
+
+  /* Setup ahafs */
+  if ((rc = uv__setup_ahafs((const char *)absolute_path, &fd)))
```
style: please do this in 2 steps (assign rc, then check it's result)

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +889,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+  uv__io_stop(handle->loop, &handle->event_watcher, UV__POLLIN);
+  uv__handle_stop(handle);
+
+  file_is_directory = uv__file_is_a_directory(handle->path);
```
No need to use a temp variable

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +889,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+  if (!uv__is_active(handle))
+    return 0;
+
+  uv__io_stop(handle->loop, &handle->event_watcher, UV__POLLIN);
```
Use uv__io_close

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +889,102 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+
+  free(handle->path);
+  handle->path = NULL;
+  close(handle->event_watcher.fd);
```
Use uv__close

saghul commented on src/unix/internal.h:
```diff
@@ -42,6 +42,16 @@
+#include <sys/poll.h>
+#undef reqevents
+#undef rtnevents
+#undef events
```
Why undef these 2? Isn't it enough to undef requevents and rtnevents?

andrewlow replied:
it was paranoid code - not required.

saghul commented:
Finished another review round @andrewlow. Almost there!

andrewlow commented:
Interesting as we've gone through the process. 

Some of this code was initially created by a junior developer and wasn't reviewed carefully locally (I should have done a much closer review before trying the pull request). Other issues have been with coding standards / formatting [I so wish we had better tools to help here].

As we've progressed - I've been making changes as suggested, and that's then introduced new issues. For example - we used to have printf's with warnings, which I changed into asserts, that have now become error code returns. Overall the code is much cleaner (and I thank you for that).

Hopefully future pulls will be cleaner and more in line with coding standards.

saghul commented on src/unix/aix.c:
```diff
@@ -39,12 +39,235 @@
+
+void uv__platform_loop_delete(uv_loop_t* loop) {
+  if (loop->fs_fd != -1) {
+    close(loop->fs_fd);
```
Use uv__close

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +308,163 @@ int uv_exepath(char* buffer, size_t* size) {
+
+    /* Get the paths. */
+    char *path = getenv("PATH");
+    if(8192 <= strlen(path))  /* verify we fit in 8k buffer for PATH */
```
sizeof(clonedpath), just in case we ever change the size

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +308,163 @@ int uv_exepath(char* buffer, size_t* size) {
+    /* Get the paths. */
+    char *path = getenv("PATH");
+    if(8192 <= strlen(path))  /* verify we fit in 8k buffer for PATH */
+      return -EINVAL;
```
Here we return an error in case path is too long, but we have some asserts for sizes above. Maybe we can also return errors in those cases?

andrewlow replied:
This error is when the fixed size buffer being used to store the environment variable PATH is not big enough to handle the path.

The asserts are indicators of programming errors - they are checking the size passed into this function. that value really should be PATH_MAX but can be any value large enough to store the fully qualified path name to the executable.

If you feel strongly here - I'll change it, but I think it's correct as is.

saghul replied:
On 07/22/2014 05:22 PM, Andrew Low wrote:

> In src/unix/aix.c:
> 
> > -    } else {
> > -      assert(*size > (strlen(symlink) + 1 + strlen(temp_buffer)));
> > -      snprintf(buffer, *size-1, "%s/%s", dirname(symlink), temp_buffer);
> > -    }
> > -    *size = strlen(buffer);
> > -    return 0;
> > -  /\* Usage of absolute filename with location exported in PATH */
> > -  } else {
> > -    char clonedpath[8192];  /\* assume 8k buffer will fit PATH */
> > -    char *token = NULL;
> > -    struct stat statstruct;
> >   +
> > -    /\* Get the paths. */
> > -    char *path = getenv("PATH");
> > -    if(8192 <= strlen(path))  /\* verify we fit in 8k buffer for PATH */
> > -      return -EINVAL;
> 
> This error is when the fixed size buffer being used to store the
> environment variable PATH is not big enough to handle the path.
> 
> The asserts are indicators of programming errors - they are checking the
> size passed into this function. that value really should be PATH_MAX but
> can be any value large enough to store the fully qualified path name to
> the executable.

Ok.

## 

Saúl Ibarra Corretgé
bettercallsaghul.com

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +308,163 @@ int uv_exepath(char* buffer, size_t* size) {
+          /* File exists. Resolve if it is a link. */
+          res = readlink(symlink, temp_buffer, PATH_MAX-1);
+          if (res < 0) {
+            assert(*size > strlen(symlink));
```
Ditto

saghul commented on src/unix/aix.c:
```diff
@@ -89,9 +308,163 @@ int uv_exepath(char* buffer, size_t* size) {
+            assert(*size > strlen(symlink));
+            strcpy(buffer, symlink);
+          } else {
+            assert(*size > (strlen(symlink) + 1 + strlen(temp_buffer)));
```
Ditto

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,381 @@ void uv_loadavg(double avg[3]) {
+
+  p = malloc(siz);
+  if (p == NULL) {
+    assert(0 && "malloc failed (errno)");
```
remove assert, let the error propagate, since this function is called as a result of the user calling uv_fs_event_start

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,381 @@ void uv_loadavg(double avg[3]) {
+    }
+    if (rv < 0) {
+      int err = errno;
+      assert(0 && "mntctl failed (errno)");
```
Ditto

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,381 @@ void uv_loadavg(double avg[3]) {
+    p = malloc(siz);
+
+    if (p == NULL) {
+      assert(0 && "realloc failed (errno)");
```
Ditto

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,381 @@ void uv_loadavg(double avg[3]) {
+  len = strlen(tmp);
+  if (tmp[len - 1] == '/')
+    tmp[len - 1] = 0;
+  for (p = tmp + 1; *p; p++)
```
even if it's one statement, please sue braces in this for loop, its body is quite big

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,381 @@ void uv_loadavg(double avg[3]) {
+    return -1;
+
+  if (sscanf(p, "RC_FROM_EVPROD=%d\nEND_EVENT_DATA", &evp_rc) == 1) {
+
```
extra newline

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,381 @@ void uv_loadavg(double avg[3]) {
+  for(i = 0; i < PATH_MAX; i++) {
+    fname[i] = 0;
+  }
+  i=0;
```
style: spaces

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,381 @@ void uv_loadavg(double avg[3]) {
+  bytes = pread(event_watch->fd, result_data, RDWR_BUF_SIZE, 0);
+
+  assert((bytes < 0) && "uv__ahafs_event - Error reading monitor file");
+  assert((bytes == 0) && "uv__ahafs_event - Error reading monitor file");
```
compact into a single assert?

saghul commented on src/unix/aix.c:
```diff
@@ -137,17 +883,99 @@ int uv_fs_event_start(uv_fs_event_t* handle,
+    absolute_path[i] = 0;
+    fname[i] = 0;
+  }
+  i=0;
```
style: spaces

saghul commented on src/unix/internal.h:
```diff
@@ -42,6 +42,14 @@
+#include <sys/poll.h>
+#undef events
+#undef revents
+#endif /* _AIX */
```
Hum, shouldn't his undef reqevents and rtnevents instead?

andrewlow replied:
Doing so results in the code failing to compile.

On Linux we have

```
/* Data structure describing a polling request.  */
struct pollfd
  {
    int fd;                     /* File descriptor to poll.  */
    short int events;           /* Types of events poller cares about.  */
    short int revents;          /* Types of events that actually occurred.  */
  };
```

On AIX it's different

```
#ifdef  __64BIT__
/* pollfd.fd is always a 32 bit entity (XOPEN) */
struct  pollfd
{
        int     fd;                     /* file descriptor or file ptr  */
        short   events;                 /* requested events             */
        short   revents;                /* returned events              */

};
#else /* __64BIT__ */
struct  pollfd
{
        long    fd;                     /* file descriptor or file ptr  */
        ushort  reqevents;              /* requested events             */
        ushort  rtnevents;              /* returned events              */
};
#define events  reqevents               /* SVR3,4 pollfd member name    */
#define revents rtnevents               /* SVR3,4 pollfd member name    */
#endif /* __64BIT__ */
```

The code in question here is trying to 'undo' the AIX-ism to make it more like Linux.

It's probably safe to remove the remain undef's (it appears to compile and test ok)

saghul commented:
I did one more round. Very few things pending :-)

> Interesting as we've gone through the process.
> 
> Some of this code was initially created by a junior developer and wasn't reviewed carefully locally (I should have done a much closer review before trying the pull request). Other issues have been with coding standards / formatting [I so wish we had better tools to help here].
> 
> As we've progressed - I've been making changes as suggested, and that's then introduced new issues. For example - we used to have printf's with warnings, which I changed into asserts, that have now become error code returns. Overall the code is much cleaner (and I thank you for that).
> 
> Hopefully future pulls will be cleaner and more in line with coding standards.

It's great to be able to colaborate with talented developers towards a common goal, I'm very thanksful for your dedication and responsiveness when handling this PR.

andrewlow commented:
After thinking about it a bit I was on the fence about returning an error code vs. an assert. The assert is much more useful to trapping incorrect usage and pointing to the problem than getting an error code back. Appreciate you're being a bit flexible on the point.

I found a couple more missing spaces around assignment using the following grep

```
grep -n =[^\ ] * | grep -v ==
```

This did turn up a couple of other files that are wrong (not part of my pull request)

```
src/unix/android-ifaddrs.c:563:        for(i=0; i<(l_prefix/8); ++i)
src/unix/freebsd.c:311:    cur+=CPUSTATES;
src/unix/tcp.c:288:    err =uv__tcp_keepalive(uv__stream_fd(handle), on, delay);
```

saghul commented on src/unix/aix.c:
```diff
@@ -128,8 +501,375 @@ void uv_loadavg(double avg[3]) {
+    }
+
+    /* Use a bigger size buffer to hold all mounted filesystems */
+    siz = sizeof(struct vmount) * size_multiplier * i;
```
I have no experience with this, but according to http://www-01.ibm.com/support/knowledgecenter/ssw_aix_53/com.ibm.aix.basetechref/doc/basetrf1/mntctl.htm%23isl60marc

"If the Size parameter indicates the supplied buffer is too small to hold the vmount structures for all the current VFSs, the mntctl subroutine sets the first word of the Buffer parameter to the required size (in bytes) and returns the value 0"

So, can we use that instead of the infinite loop and the malloc multiplier?

saghul commented:
Left one last comment abount `mnctl`, otherwise this is good to land :-)

andrewlow commented:
`mntctl` comment was correct -- code fixed.

saghul commented:
That was quite a PR journey :-) Great work, Andrew! Landed in :sparkles: https://github.com/joyent/libuv/commit/cdc979dbe04bac58eb33ee2510f860c172a42d7e :sparkles:
